### PR TITLE
feat(build): #1094 update formatPython

### LINKED ISSUE #1094: `FormatPython` fails because does not support new  Python syntax
# Problem to solve
New Python syntax is not supported.

# Intended users
Alls.

# Proposal
Upgrade target-version in [settings-black.toml](https://github.com/fluidattacks/makes/blob/9c3b8a1f38d875674c2ae7a5a3b02b62d2993216/src/evaluator/modules/format-python/settings-black.toml)

Note: [Which Python versions does Black support?](https://black.readthedocs.io/en/stable/faq.html#which-python-versions-does-black-support)

### PR DESCRIPTION
**Pull request by rohaquinlop**
Fix #1094 
- Update formatPython package to use python 3.11
- Separate the arg from evaluator
- Allow having several configurations
- Allow passing isort and black settings to each configuration
- Add job to dev pipeline
- Update documentation

### CODE REVIEW

**Comment by dsalaza4** [comment]
Hi @rohaquinlop,

First of all, thank you for this contribution!

Now that we're dealing with `formatPython`, maybe we should:

1. Separate the arg from the evaluator.
2. Allow having several configurations.
3. Allow passing isort and black settings to each configuration.

Ideally, it would look like this:

```nix
# makes.nix
{
  formatPython = {
    strict = {
      enabled = true;
      config = {
        black = ./config/black/strict.toml;
        isort = ./config/isort/strict.toml;
      };
    };
    flexible = {
      enabled = true;
      config = {
        black = ./config/black/flexible.toml;
      };
    };
  };
}
```

You can use other builtins like `testPullRequest` [arg](https://github.com/fluidattacks/makes/tree/main/src/args/test-pull-request) and [evaluator](https://github.com/fluidattacks/makes/tree/main/src/evaluator/modules/test-pull-request) as a reference for achieving all of this.

**Comment by rohaquinlop** [comment]
@dsalaza4 Got it! I'm gonna work on it.

One question, is there any documentation to build from source?

**Comment by dsalaza4** [comment]
> @dsalaza4 Got it! I'm gonna work on it.
> 
> One question, is there any documentation to build from source?

Just [install](https://makes.fluidattacks.com/getting-started/#installation) and run `m . <job>` from your local repo, that should do the trick.

**Comment by dsalaza4** [comment]
@rohaquinlop Let's get rid of the `enabled` argument, it is unnecessary.

**Comment by rohaquinlop** [comment]
@dsalaza4 I have one question, there will be only these two configurations? `strict` and `flexible`? Because I've been trying to implement it allowing different config names, but it wasn't successfully.

Here's my current implementation allowing different configuration names:

```nix
{
  __toModuleOutputs__,
  formatPython,
  ...
}: {
  config,
  lib,
  ...
}: let
  makeOutput = name: cfg: {
    settingsBlack,
    settingsIsort,
    targets,
  }: {
    name = "/formatPython/${name}";
    value = formatPython {
      inherit settingsBlack;
      inherit settingsIsort;
      inherit targets;
      inherit name;
    };
  };
in {
  options = {
    formatPython = {
      configurations = lib.mkOption {
        type = lib.types.attrsOf (lib.types.submodule (_: {
          options = {
            settingsBlack = lib.mkOption {
              default = ./settings-black.toml;
              type = lib.types.path;
            };
            settingsIsort = lib.mkOption {
              default = ./settings-isort.toml;
              type = lib.types.path;
            };
            targets = lib.mkOption {
              default = ["/"];
              type = lib.types.listOf lib.types.str;
            };
          };
        }));
      };
    };
  };
  config = {
    outputs = __toModuleOutputs__ (lib.mapAttrs' makeOutput config.formatPython.configurations);
  };
}
```

And I'm getting the following error:
```
error: The option `formatPython.strict' does not exist. Definition values:
       - In `<unknown-file>':
           {
             config = {
               targets = [
                 "/"
               ];
           ...
```

```nix
# makes.nix
{
  formatPython = {
    strict = {
      config = {
        targets = ["/"];
      };
    };
  };
}
```

**Comment by dsalaza4** [comment]
@rohaquinlop `strict` and `flexible` are not static configurations but rather example names for configurations that receive a given set of configs. You can create as many configurations as you want within the `formatPython` block.

```nix
# makes.nix
{
  formatPython = {
    my-config-1 = {
      enabled = true;
      config = {
        black = ./path-in-my-repo-to-a-config.toml;
        isort = ./path-in-my-repo-to-another-config.toml;
      };
    };
    my-config-2 = {
      enabled = true;
      # No configs provided, use default ones
    };
  };
}
```

**Comment by dsalaza4** [comment]
Regarding the error you're getting, I think the reason is that you have a `confiugurations` block within `formatPython` that is receiving the configs. Try using `formatPython.configurations.strict`. In any case, I think we should remove the `configurations` block and stick to the `formatPython.strict` format.

**Comment by rohaquinlop** [comment]
@dsalaza4 I'm getting an error and I can't figure out why is it.

Error:
```
error: getting status of '/nix/store/yd084mqsjqlj6n5wfwmv10x3gnq84p78-makes-ac32mpkr/src/args/format-python/default.nix': No such file or directory
```

But I already added formatPython to the `agnostic.nix` file in the path `/src/args/` do you have any idea why I'm getting this error?

**Comment by dsalaza4** [comment]
Maybe you're not adding the new files to the repository. Try at least adding all new files to staging with a `git add`.

File: makes.nix [hunk 1]
```diff
@@ -119,8 +119,13 @@
     targets = ["/"];
   };
   formatPython = {
-    enable = true;
-    targets = ["/"];
+    example = {
+      config = {
+        targets = ["/"];
```
Comment (dsalaza4) [inline makes.nix:124]: Let's move targets outside of config:

```nix
{
  formatPython = {
    example = {
      targets = ["/"];
      config = {...};
    };
  };
}
```

File: makes.nix [hunk 2]
```diff
@@ -119,8 +119,13 @@
     targets = ["/"];
   };
   formatPython = {
-    enable = true;
-    targets = ["/"];
+    example = {
```
Comment (dsalaza4) [inline makes.nix:122]: Let's call it `default` instead of `example`.

File: makes.nix [hunk 3]
```diff
@@ -119,8 +119,13 @@
...
+      config = {
+        targets = ["/"];
+      };
+    };
+    my-config = {
```
Comment (dsalaza4) [inline makes.nix:127]: Remove this.

File: src/args/format-python/default.nix [hunk 4]
```diff
@@ -0,0 +1,30 @@
+{
+  __nixpkgs__,
+  toBashArray,
+  makeScript,
+  ...
+}: {
+  settingsBlack,
```
Comment (dsalaza4) [inline src/args/format-python/default.nix:7]: Receive the settings map:

```
    __argSettingsBlack__ = settings.black;
    __argSettingsIsort__ = settings.isort;
```

File: src/evaluator/modules/format-python/default.nix [hunk 5]
```diff
@@ -1,48 +1,46 @@
...
-            __nixpkgs__.python39Packages.colorama
-          ];
+in {
+  options = {
+    formatPython = lib.mkOption {
+      default = {};
+      type = lib.types.attrsOf (lib.types.submodule (_: {
+        options = {
+          config = {
+            settingsBlack = lib.mkOption {
```
Comment (dsalaza4) [inline src/evaluator/modules/format-python/default.nix:26]:
```suggestion
            black = lib.mkOption {
```

File: src/evaluator/modules/format-python/default.nix [hunk 6]
```diff
@@ -1,48 +1,46 @@
...
+              default = ./settings-black.toml;
+              type = lib.types.path;
+            };
+            settingsIsort = lib.mkOption {
```
Comment (dsalaza4) [inline src/evaluator/modules/format-python/default.nix:30]:
```suggestion
            isort = lib.mkOption {
```

File: makes.nix [hunk 7]
```diff
@@ -119,8 +119,13 @@
...
+    };
   };
   formatTerraform = {
     enable = true;
```
Comment (dsalaza4) [inline makes.nix:131]: Unnecessary?

File: .github/workflows/dev.yml [hunk 8]
```diff
@@ -131,17 +131,17 @@ jobs:
     steps:
       - uses: actions/checkout@f095bcc56b7c2baf48f3ac70d6d6782f4f553222
       - uses: docker://docker.io/nixos/nix@sha256:c3db4c484f6b1ee6c9bb8ca90307cfbeca8ef88156840911356a677eeaff4845
-        name: /formatPython
+        name: /formatPython/example
```
Comment (dsalaza4) [inline .github/workflows/dev.yml:134]:
```suggestion
        name: /formatPython/default
```

File: .github/workflows/dev.yml [hunk 9]
```diff
@@ -131,17 +131,17 @@ jobs:
...
         with:
           set-safe-directory: /github/workspace
-          args: sh -c "nix-env -if . && m . /formatPython"
+          args: sh -c "nix-env -if . && m . /formatPython/example"
```
Comment (dsalaza4) [inline .github/workflows/dev.yml:137]:
```suggestion
          args: sh -c "nix-env -if . && m . /formatPython/default"
```

File: .github/workflows/dev.yml [hunk 10]
```diff
@@ -131,17 +131,17 @@ jobs:
...
   macos_formatPython:
     runs-on: macos-latest
     steps:
       - uses: actions/checkout@f095bcc56b7c2baf48f3ac70d6d6782f4f553222
       - uses: cachix/install-nix-action@67e9fd765dbe63fabe0ce2bd72f56f9a417696a0
-      - name: /formatPython
-        run: nix-env -if . && m . /formatPython
+      - name: /formatPython/example
```
Comment (dsalaza4) [inline .github/workflows/dev.yml:143]:
```suggestion
      - name: /formatPython/default
```

File: .github/workflows/dev.yml [hunk 11]
```diff
@@ -131,17 +131,17 @@ jobs:
...
+        run: nix-env -if . && m . /formatPython/example
```
Comment (dsalaza4) [inline .github/workflows/dev.yml:144]:
```suggestion
        run: nix-env -if . && m . /formatPython/default
```

File: .github/workflows/dev.yml (hunk 8 above)
Comment (rohaquinlop) [reply .github/workflows/dev.yml:134]: My bad, I forget to update it.

**Comment by dsalaza4** [comment]
Please update documentation accordingly: https://makes.fluidattacks.com/api/builtins/format/?h=formatp#formatpython

You can use `m . /docs/dev` for a local rendering.

File: docs/src/api/builtins/format.md [hunk 12]
```diff
@@ -136,20 +144,18 @@ Example:
...
       formatPython = {
-        enable = true;
-        targets = [
-          "/" # Entire project
-          "/file.py" # A file
-          "/directory" # A directory within the project
-        ];
+        my-config = {
+          targets = ["/"];
+          config = {...};
```
Comment (dsalaza4) [inline docs/src/api/builtins/format.md:149]: This is invalid nix. Let’s provide valid code blocks.
```suggestion
          config = {};
```

**Comment by dsalaza4** [comment]
Please add yourself to the mailmap: https://github.com/fluidattacks/makes/actions/runs/5245872757/jobs/9503238429?pr=1106

**Comment by rohaquinlop** [comment]
> Please add yourself to the [mailmap](https://github.com/fluidattacks/makes/blob/main/.mailmap): https://github.com/fluidattacks/makes/actions/runs/5245872757/jobs/9503238429?pr=1106

Ok, I'm gonna add me

**Comment by rohaquinlop** [comment]
> > > Please add yourself to the [mailmap](https://github.com/fluidattacks/makes/blob/main/.mailmap): https://github.com/fluidattacks/makes/actions/runs/5245872757/jobs/9503238429?pr=1106
> > 
> > 
> > Ok, I'm gonna add me
> 
> Still failing? You can test with `m . /lintGitMailMap`

Thank you, didn't update the author's commit.

**Review by dsalaza4 — APPROVED** [review]
LGTM!

Thank you so much for all your work!

**Comment by rohaquinlop** [comment]
> LGTM!
> 
> Thank you so much for all your work!

Thank you to, your quick responses and your patience with me

**Pull request merged**